logger/memlogger/memlogger.go

```diff
@@ -0,0 +1,234 @@
+	ms.log("DEBUG", template, args, nil, nil)
+}
+
+func (ms *Store) Infof(template string, args ...any) {
+	ms.log("INFO", template, args, nil, nil)
+}
+
+func (ms *Store) Warnf(template string, args ...any) {
+	ms.log("WARN", template, args, nil, nil)
+}
+
+func (ms *Store) Errorf(template string, args ...any) {
+	ms.log("ERROR", template, args, nil, nil)
+}
+
+func (ms *Store) Fatalf(template string, args ...any) {
+	ms.log("FATAL", template, args, nil, nil)
+}
+
+func (ms *Store) Debugw(msg string, keysAndValues ...any) {
+	ms.log("DEBUG", msg, nil, keysAndValues, nil)
+}
+
+func (ms *Store) Infow(msg string, keysAndValues ...any) {
+	ms.log("INFO", msg, nil, keysAndValues, nil)
+}
+
+func (ms *Store) Warnw(msg string, keysAndValues ...any) {
+	ms.log("WARN", msg, nil, keysAndValues, nil)
+}
+
+func (ms *Store) Errorw(msg string, keysAndValues ...any) {
+	ms.log("ERROR", msg, nil, keysAndValues, nil)
+}
+
+func (ms *Store) Fatalw(msg string, keysAndValues ...any) {
+	ms.log("FATAL", msg, nil, keysAndValues, nil)
+}
+
+func (ms *Store) Debugn(msg string, fields ...logger.Field) {
+	ms.log("DEBUG", msg, nil, nil, fields)
+}
+
+func (ms *Store) Infon(msg string, fields ...logger.Field) {
+	ms.log("INFO", msg, nil, nil, fields)
+}
+
+func (ms *Store) Warnn(msg string, fields ...logger.Field) {
+	ms.log("WARN", msg, nil, nil, fields)
+}
+
+func (ms *Store) Errorn(msg string, fields ...logger.Field) {
+	ms.log("ERROR", msg, nil, nil, fields)
+}
+
+func (ms *Store) Fataln(format string, fields ...logger.Field) {
+	ms.log("FATAL", format, nil, nil, fields)
+}
+
+func (ms *Store) Child(s string) logger.Logger {
+	ms.mu.Lock()
+	defer ms.mu.Unlock()
+
+	child := New()
+	child.commonKeysAndValues = append(child.commonKeysAndValues, ms.commonKeysAndValues...)
+	child.commonFields = append(child.commonFields, ms.commonFields...)
+
+	if ms.name == "" {
+		child.name = s
+	} else {
+		child.name = strings.Join([]string{ms.name, s}, ".")
+	}
+
+	ms.childStores = append(ms.childStores, child)
+	return child
+}
+
+func (ms *Store) With(args ...any) logger.Logger {
+	ms.mu.Lock()
+	defer ms.mu.Unlock()
+
+	ms.commonKeysAndValues = append(ms.commonKeysAndValues, args...)
+	return ms
+}
+
+func (ms *Store) Withn(args ...logger.Field) logger.Logger {
+	ms.mu.Lock()
+	defer ms.mu.Unlock()
+
+	ms.commonFields = append(ms.commonFields, args...)
+	return ms
+}
+
+func (ms *Store) log(level, template string, fmtArgs, keysAndValues []interface{}, fields []logger.Field) {
+	ms.mu.Lock()
+	defer ms.mu.Unlock()
+
+	ms.entries = append(ms.entries, entry{
+		level:         level,
+		template:      template,
+		fmtArgs:       fmtArgs,
+		keysAndValues: append(ms.commonKeysAndValues, keysAndValues...),
+		fields:        append(ms.commonFields, fields...),
+	})
+}
+
+type SearchOptions struct {
+	Name          string
+	Level         string
+	Msg           string
+	KeysAndValues []interface{}
+	Fields        []logger.Field
+}
+
+func (ms *Store) Search(opts SearchOptions) int {
```

Comment:
I am not sure about this method returning an int.
IMHO it would be more appropriate and convenient for the name for Search to return a slice of log entries. 

I image the following usage:
```
foo.Bar()

require.Equal(t, logStore.Search("http request failed"), []memlog.Log{{
   Level: "Debug",
   Message: "http request failed", 
   KeysAndValues: nil, 
}})
```

In the example above it will be much easier to understand that the level or KeysAndValues have changed. Instead of getting zero values in a search, you will get the diff for the log.